### PR TITLE
Use trackErrors ce instead of ++ custom operator

### PR DESCRIPTION
**Pull request by edgarfgp**
While contributing I always find confusing using custom operators.

In this PR my goal is to replace the usage of 
```fsharp
val (++): res: OperationResult<'T> -> f: ('T -> OperationResult<'b>) -> OperationResult<'b>`
```

with

```fsharp
trackErrors {
    ....
}

### CODE REVIEW

File: src/Compiler/Checking/AttributeChecking.fs [hunk 1]
```diff
@@ -266,29 +266,28 @@ let langVersionPrefix = "--langversion:preview"
...
+let CheckFSharpAttributes (g:TcGlobals) attribs m = trackErrors {
     let isExperimentalAttributeDisabled (s:string) =
         if g.compilingFSharpCore then
             true
         else
             g.langVersion.IsPreviewEnabled && (s.IndexOf(langVersionPrefix, StringComparison.OrdinalIgnoreCase) >= 0)
 
-    if isNil attribs then CompleteD
+    if isNil attribs then do! CompleteD
     else
```
Comment (kerams) [inline src/Compiler/Checking/AttributeChecking.fs:277]: Can the builder be moved to the `else` branch? That way we don't incur any overhead on the main path (no attributes).

Comment (edgarfgp) [reply src/Compiler/Checking/AttributeChecking.fs:271]: Yeah, I think so. Thanks for the review :)

**Comment by edgarfgp** [comment]
I am not entirely sure how to convert the last one. 

```fsharp
TryD (fun () -> SolveTypeSubsumesType csenv 0 m NoTrace None reqdObjTy availObjTy ++ (fun () -> ResultD true))
                 (fun _err -> ResultD false)
```

https://github.com/dotnet/fsharp/blob/0514e881c2e03b2c004504b8b21432aa4a01c7bc/src/Compiler/Checking/ConstraintSolver.fs#L3769C94-L3769C100

File: src/Compiler/Checking/AttributeChecking.fs [hunk 2]
```diff
@@ -503,15 +503,17 @@ let PropInfoIsUnseen m pinfo =
 #endif
 
 /// Check the attributes on a union case, returning errors and warnings as data.
-let CheckUnionCaseAttributes g (x:UnionCaseRef) m =
-    CheckEntityAttributes g x.TyconRef m ++ (fun () ->
-    CheckFSharpAttributes g x.Attribs m)
+let CheckUnionCaseAttributes g (x:UnionCaseRef) m = trackErrors {
+    do! CheckEntityAttributes g x.TyconRef m
+    do! CheckFSharpAttributes g x.Attribs m
+}
```
Comment (vzarytovskii) [inline src/Compiler/Checking/AttributeChecking.fs:509]: Could you please re-format it, to have the CE on the new line, it's arguably easier to understand that you're inside specific CE when it's visible on its own line?

**Comment by vzarytovskii** [comment]
That's good, we can probably remove the `(++)` if it's not used anywhere. I find its usage confusing and not useful. Thanks for rewriting it.

**Comment by edgarfgp** [comment]
> That's good, we can probably remove the `(++)` if it's not used anywhere. I find its usage confusing and not useful. Thanks for rewriting it.

 `(++)` is now only used in `DiagnosticsLogger`  `TrackErrorsBuilder`  and in some of its helper functions. So I guess We can rewrite it and remove the usage of the custom operator In a separate PR ?

**Review by psfinaki — APPROVED** [review]
Thanks for this.

As for removing (++) completely - as usual, I'd prefer a separate PR but not opposed to having it here either :)

**Comment by edgarfgp** [comment]
> I'm looking back at this, there are a bunch of `do! CompleteD`. Isn't that essentially a no-op?

Do you mean that these can go outside of the `trackErrors` ce ?

**Comment by kerams** [comment]
Does it do anything? It seems to me it could be replaced with `()` and some branches could be eliminated altogether.

**Comment by edgarfgp** [comment]
> Does it do anything? It seems to me it could be replaced with `()` and some branches could be eliminated altogether.

Yeah that's a good point. On this PR I wanted to just replace the ++ operator, But I think it can be optimized in follow-up PR

**Pull request merged**